incubator/virtualcluster/pkg/syncer/resources/ingress/controller.go

```diff
@@ -0,0 +1,178 @@
+	}
+	var mcOptions *mc.Options
+	if options == nil || options.MCOptions == nil {
+		mcOptions = &mc.Options{Reconciler: c}
+	} else {
+		mcOptions = options.MCOptions
+	}
+	mcOptions.MaxConcurrentReconciles = constants.DwsControllerWorkerLow
+	multiClusterIngressController, err := mc.NewMCController("tenant-masters-ingress-controller", &v1beta1.Ingress{}, *mcOptions)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to create ingress mc controller: %v", err)
+	}
+	c.multiClusterIngressController = multiClusterIngressController
+
+	c.ingressLister = informer.Extensions().V1beta1().Ingresses().Lister()
+	if options != nil && options.IsFake {
+		c.ingressSynced = func() bool { return true }
+	} else {
+		c.ingressSynced = informer.Extensions().V1beta1().Ingresses().Informer().HasSynced
+	}
+
+	var uwOptions *uw.Options
+	if options == nil || options.UWOptions == nil {
+		uwOptions = &uw.Options{Reconciler: c}
+	} else {
+		uwOptions = options.UWOptions
+	}
+	uwOptions.MaxConcurrentReconciles = constants.UwsControllerWorkerLow
+	upwardIngressController, err := uw.NewUWController("ingress-upward-controller", &v1beta1.Ingress{}, *uwOptions)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to create ingress upward controller: %v", err)
+	}
+	c.upwardIngressController = upwardIngressController
+
+	var patrolOptions *pa.Options
+	if options == nil || options.PatrolOptions == nil {
+		patrolOptions = &pa.Options{Reconciler: c}
+	} else {
+		patrolOptions = options.PatrolOptions
+	}
+	ingressPatroller, err := pa.NewPatroller("ingress-patroller", &v1beta1.Ingress{}, *patrolOptions)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to create ingress patroller: %v", err)
+	}
+	c.ingressPatroller = ingressPatroller
+
+	informer.Extensions().V1beta1().Ingresses().Informer().AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1beta1.Ingress:
+					return isBackPopulateIngress(t)
+				case cache.DeletedFinalStateUnknown:
+					if e, ok := t.Obj.(*v1beta1.Ingress); ok {
+						return isBackPopulateIngress(e)
+					}
+					utilruntime.HandleError(fmt.Errorf("unable to convert object %v to *v1beta1.Ingress", obj))
+					return false
+				default:
+					utilruntime.HandleError(fmt.Errorf("unable to handle object in super master ingress controller: %v", obj))
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc: c.enqueueIngress,
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					newIngress := newObj.(*v1beta1.Ingress)
+					oldIngress := oldObj.(*v1beta1.Ingress)
+					if newIngress.ResourceVersion != oldIngress.ResourceVersion {
+						c.enqueueIngress(newObj)
+					}
+				},
+				DeleteFunc: c.enqueueIngress,
+			},
+		})
+	return c, multiClusterIngressController, upwardIngressController, nil
+}
+
+func isBackPopulateIngress(ing *v1beta1.Ingress) bool {
```

Comment:
Should not this method get removed? I think we should always backpopulate ingress for status update.